src/mbgl/renderer/layers/render_location_indicator_layer.cpp

```diff
@@ -0,0 +1,806 @@
+
+    struct Shader {
+        virtual ~Shader() { release(); }
+        void release() {
+            if (!program) return;
+            MBGL_CHECK_ERROR(glDetachShader(program, vertexShader));
+            MBGL_CHECK_ERROR(glDetachShader(program, fragmentShader));
+            MBGL_CHECK_ERROR(glDeleteShader(vertexShader));
+            MBGL_CHECK_ERROR(glDeleteShader(fragmentShader));
+            MBGL_CHECK_ERROR(glDeleteProgram(program));
+            program = vertexShader = fragmentShader = 0;
+        }
+        void initialize(const GLchar* const& vsSource, const GLchar* const& fsSource) {
+            if (program) return;
+            program = MBGL_CHECK_ERROR(glCreateProgram());
+            vertexShader = MBGL_CHECK_ERROR(glCreateShader(GL_VERTEX_SHADER));
+            fragmentShader = MBGL_CHECK_ERROR(glCreateShader(GL_FRAGMENT_SHADER));
+            MBGL_CHECK_ERROR(glShaderSource(vertexShader, 1, &vsSource, nullptr));
+            MBGL_CHECK_ERROR(glCompileShader(vertexShader));
+            MBGL_CHECK_ERROR(glAttachShader(program, vertexShader));
+            MBGL_CHECK_ERROR(glShaderSource(fragmentShader, 1, &fsSource, nullptr));
+            MBGL_CHECK_ERROR(glCompileShader(fragmentShader));
+            MBGL_CHECK_ERROR(glAttachShader(program, fragmentShader));
+            MBGL_CHECK_ERROR(glLinkProgram(program));
+            pullLocations();
+        }
+        virtual void bind() { MBGL_CHECK_ERROR(glUseProgram(program)); }
+        void detach() { MBGL_CHECK_ERROR(glUseProgram(0)); }
+        virtual void pullLocations(){};
+
+        GLuint program = 0;
+        GLuint vertexShader = 0;
+        GLuint fragmentShader = 0;
+    };
+
+    struct SimpleShader : public Shader {
+        // Note that custom layers need to draw geometry with a z value of 1 to take advantage of
+        // depth-based fragment culling.
+        const GLchar* vertexShaderSource = R"MBGL_SHADER(
+attribute vec2 a_pos;
+uniform mat4 u_matrix;
+void main() {
+    gl_Position = u_matrix * vec4(a_pos, 0, 1);
+//            gl_Position = vec4(a_pos, 0, 1);
+}
+)MBGL_SHADER";
+
+        const GLchar* fragmentShaderSource = R"MBGL_SHADER(
+uniform vec4 u_color;
+void main() {
+    gl_FragColor = u_color;
+}
+)MBGL_SHADER";
+
+        void initialize() { Shader::initialize(SimpleShader::vertexShaderSource, SimpleShader::fragmentShaderSource); }
+
+        void pullLocations() override {
+            a_pos = MBGL_CHECK_ERROR(glGetAttribLocation(program, "a_pos"));
+            u_color = MBGL_CHECK_ERROR(glGetUniformLocation(program, "u_color"));
+            u_matrix = MBGL_CHECK_ERROR(glGetUniformLocation(program, "u_matrix"));
+        }
+        void bind() override {
+            SimpleShader::initialize();
+            Shader::bind();
+        }
+
+        GLuint a_pos = 0;
+        GLuint u_color = 0;
+        GLuint u_matrix = 0;
+    };
+
+    struct TexturedShader : public Shader {
+        const GLchar* vertexShaderSource = R"MBGL_SHADER(
```

Comment:
This file should not contain GLSL code or any calls to OpenGL. Please use the `gfx` interface. Shaders are in https://github.com/mapbox/mapbox-gl-js/tree/master/src/shaders and are converted by the [`generate-shaders.js`](https://github.com/mapbox/mapbox-gl-native/blob/master/scripts/generate-shaders.js) script.

Reply:
The current idea is to use this approach, and start refactoring it into gfx as soon as it's merged